### PR TITLE
fix Issue 19607 - [ICE] dmd/e2ir.d(117): Invalid type mTYconst|TYstruct

### PR DESCRIPTION
**Pull request by ibuclaw**
Added new class to handle taking the array of a vector, similar DelegatePtrExp/DelegateFuncptrExp.  How this operation is done would differ between compilers anyway, and once again, dmd type painting expressions in the front-end results in garbage AST.

Back-end bit is just copy-paste guesswork, based on static array literal code.

### CODE REVIEW

**Comment by thewilsonator** [comment]
Does this need an entry in `UnionExp`?

**Comment by ibuclaw** [comment]
According to the one fail test change, looks like it would be better to break out the expansion of `__vector(1)` to `__vector([1, ...])` into its own function, and call it lazily where required.

**Comment by thewilsonator** [comment]
Ping me when this is good to go.

**Comment by ibuclaw** [comment]
@thewilsonator - the implementation is complete at least...

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/expression.d [hunk 1]
```diff
@@ -5094,6 +5097,30 @@ extern (C++) final class VectorExp : UnaExp
     }
 }
 
+extern (C++) final class VectorArrayExp : UnaExp
```
Comment (WalterBright) [inline src/dmd/expression.d]: A comment describing what this is would be useful, as there's no way to tell from the following code.

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -5178,6 +5172,85 @@ public:
         result = pue.exp();
     }
 
+    /**
+     * Interpret the vector expression as an array literal.
+     * Params:
+     *    pue = non-null pointer to temporary storage that can be used to store the return value
+     *    e = Expression to interpret
+     * Returns:
+     *    resulting expression
```
Comment (WalterBright) [inline src/dmd/dinterpret.d]: Say it's returning an array literal.

**Review by WalterBright — DISMISSED** [review]
Need more explanatory comments as noted.

**Comment by ibuclaw** [comment]
Rebased and added comments.

**Comment by ibuclaw** [comment]
Fixed typo in rebase...  should all be green now.

**Pull request merged**